### PR TITLE
fix: tool description builder

### PR DESCRIPTION
**Pull request by angrybayblade**
# 🔍 Review Summary 
 ### Release Note

**Purpose**:
- Enhance consistency and readability of tool descriptions by refining formatting.

**Changes**:
- **Enhancement**: Updated `set_metadata` in `abs.py` to use `_get_description` for better tool description formatting.
- **Test**: Added `test_description_builder` to validate description construction from docstrings.

**Impact**:
- Ensures consistently formatted tool descriptions, enhancing readability and user comprehension of tool functionalities. 

<details><summary>Original Description</summary>

No existing description found

</details>

----

> [!IMPORTANT]
> Improves tool description formatting in `abs.py` by using a new `_get_description` method and adds a test to validate this change.
> 
>   - **Behavior**:
>     - Updated `set_metadata` in `ToolBuilder` in `abs.py` to use `_get_description` for consistent tool description formatting.
>     - Added `test_description_builder` in `test_abs.py` to validate description construction from docstrings.
>   - **Functions**:
>     - Added `_get_description` method in `ToolBuilder` to format tool descriptions by joining non-empty lines from docstrings.
>   - **Tests**:
>     - Ensures `SomeTool.description` is correctly formatted in `test_description_builder`.
> 
> <sup>This description was created by </sup>[Ellipsis](https://www.ellipsis.dev?ref=ComposioHQ%2Fcomposio&utm_source=github&utm_medium=referral)<sup> for b50bed94b2ae7ef7b08738355073a1bc03b53a11. It will automatically update as commits are pushed.</sup>

### CODE REVIEW

File: python/composio/tools/base/abs.py [hunk 1]
```diff
@@ -426,6 +426,18 @@ def setup_children(obj: t.Type["Tool"]) -> None:
             if hasattr(obj, "logo"):
                 setattr(trigger, "logo", getattr(obj, "logo"))
 
+    @staticmethod
+    def _get_description(obj) -> str:
```
Comment (shreysingla11) [inline python/composio/tools/base/abs.py:430]: The `_get_description` method should include type hints for its parameters and a docstring explaining its functionality:

```python
@staticmethod
def _get_description(obj: t.Type["Tool"]) -> str:
    """
    Extracts and formats the description from a tool object's docstring or display name.
    
    Args:
        obj: The tool object to extract description from
        
    Returns:
        str: Formatted description with cleaned whitespace and joined multiple lines
    """
```

File: python/composio/tools/base/abs.py [hunk 2]
```diff
@@ -426,6 +426,18 @@ def setup_children(obj: t.Type["Tool"]) -> None:
...
+        return " ".join(
+            filter(
+                bool,
+                map(
+                    str.strip,
+                    (obj.__doc__ if obj.__doc__ else obj.display_name).splitlines(),
```
Comment (shreysingla11) [inline python/composio/tools/base/abs.py:436]: Consider adding error handling for cases where neither `__doc__` nor `display_name` exists to prevent potential AttributeError:

```python
description_source = obj.__doc__ if hasattr(obj, "__doc__") and obj.__doc__ else (
    obj.display_name if hasattr(obj, "display_name") else ""
)
return " ".join(filter(bool, map(str.strip, description_source.splitlines())))
```

File: python/tests/test_tools/test_base/test_abs.py [hunk 3]
```diff
@@ -153,3 +153,24 @@ def actions(cls) -> list:
         assert SomeAction.enum == "SOME_TOOL_SOME_ACTION"
         assert isinstance(tool_registry["local"][SomeTool.enum], SomeTool)
         assert action_registry["local"][SomeAction.enum] is SomeAction
+
+    def test_description_builder(self) -> None:
```
Comment (shreysingla11) [inline python/tests/test_tools/test_base/test_abs.py:157]: The test coverage could be improved by adding test cases for edge cases:
```python
def test_description_builder_edge_cases(self) -> None:
    # Test empty docstring
    class EmptyDocTool(Tool):
        pass
    
    # Test single line docstring
    class SingleLineTool(Tool):
        """Single line doc"""
        pass
    
    # Test multiple empty lines
    class MultiEmptyLineTool(Tool):
        """
        
        Some doc
        
        with spaces
        
        """
        pass
    
    # Test missing both doc and display_name
    class NoDescTool(Tool):
        pass
    setattr(NoDescTool, "__doc__", None)
```

**Comment by shreysingla11** [comment]
### Code Review Summary

#### Overall Assessment
The changes look good and improve the tool description handling by properly managing multi-line docstrings and whitespace. The code is well-structured and the addition of tests is positive.

#### Strengths
✅ Good separation of concerns with the new `_get_description` method
✅ Proper use of functional programming concepts (filter, map)
✅ Added test coverage for the new functionality
✅ Appropriate change from `@staticmethod` to `@classmethod` for `set_metadata`

#### Suggestions for Improvement
1. Add type hints and docstring for the `_get_description` method
2. Include error handling for missing attributes
3. Expand test coverage to include edge cases

The changes are ready to be merged after addressing the minor suggestions above. Good work on improving the description handling! 👍

Code Quality Rating: 8/10

**Review by tushar-composio — APPROVED** [review]
